OSDOCS-2891: Adding docs for using the oc-mirror plug-in to mirror im…

bergerhoffer opened the pull request:
…ages for disconnected installs

https://issues.redhat.com/browse/OSDOCS-2891

Previews: 
* https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/index.html
* https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-creating-registry.html
* https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-installation-images.html
* https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html

bergerhoffer commented on modules/oc-mirror-creating-image-set-config.adoc:
```diff
@@ -0,0 +1,63 @@
+
+// Prior to running `oc mirror` / mirroring images using the `oc-mirror` plug-in, you must create the `ImageSetConfiguration` resource to define... TODO.
+
+// TODO Need more information on how this config works. Do you use the same exact config both when doing mirror2disk and disk2mirror?
```
@afflom @jpower432 Do we expect users to use the same exact imageset-config.yaml file both when running the mirror2disk and the disk2mirror method? So they use the file to download the image set to disk, then transfer that config file to the disconnected env as well and then use it there to do disk2mirror? Or would the image set config file need to be different?

jpower432 replied:
On the disconnected side, the input would be the generated imageset. disk2mirror does not accept a config file. So instead of the flag being `--config imageset-config.yaml`, in disk2mirror operations, it would be `--from mirror_seq1_00000.tar`. It might help this think of the command as `oc mirror prefix://<destination>` with the `config` and `from` flags representing the user input artifact, be it a config or imageset. Does this help?

bergerhoffer replied:
Jeeze you're right, I'm not sure why I didn't notice that we weren't passing in the image set config file here! Thanks, I'll update everywhere that I made that assumption.

bergerhoffer commented on modules/oc-mirror-creating-image-set-config.adoc:
```diff
@@ -0,0 +1,63 @@
+<5> URL to query for chart information.
+<6> Helm chart to add to the image set.
+<7> Optional: Chart version to add to the image set.
+<8> Optional: Tag of image to add to the image set.
```
@afflom @jpower432 The description for this in the Google doc said "Tag of optional image to be added to the imageset". I wasn't sure what "optional image" meant here. Or what exactly is optional here

afflom replied:
If the user needs arbitrary images not contained within an ocp release, operator catalog, or helm chart, those images can be annotated here.

bergerhoffer commented on modules/oc-mirror-creating-image-set-config.adoc:
```diff
@@ -0,0 +1,63 @@
+<6> Helm chart to add to the image set.
+<7> Optional: Chart version to add to the image set.
+<8> Optional: Tag of image to add to the image set.
+<9> Optional: Back-end location to save the image set metadata to.
```
@afflom @jpower432 We (Alex and I) had talked about only supporting the stateful method in the product docs, which I think meant that the `storageConfig` section is required. Is that true, and if so, we should probably state that here somehow.

bergerhoffer commented on modules/oc-mirror-disk-to-mirror.adoc:
```diff
@@ -0,0 +1,57 @@
+* You have transferred the image set file and the image set configuration file to the disconnected environment.
+// TODO: Confirm prereq about not needing a cluster, but need pull secret misc
+
+// TODO: confirm that they needed to transfer the same image set config file to the disconnected env
```
@afflom @jpower432 Can you confirm that we do expect people to transfer that imageset-config.yaml file to the disconnected environment? Or somehow get a similar/the same file there somehow?

bergerhoffer commented on modules/oc-mirror-disk-to-mirror.adoc:
```diff
@@ -0,0 +1,57 @@
+<2> Specify the registry to mirror the image set file to. The registry must start with `docker://`.
++
+// TODO: for <1>: Can there be more than one .tar file in the directory?
+// TODO: for <2>: Confirm that the docker:// start is an actual requirement.
```
@afflom @jpower432 The Google doc mentioned that the `docker://` prefix was required. Just checking that this is true (I have a statement above that says it must start with that)

bergerhoffer commented on modules/oc-mirror-disk-to-mirror.adoc:
```diff
@@ -0,0 +1,57 @@
+<1> Pass in the image set .tar file to mirror. This procedure assumes that it is named `mirror_seq1_000000.tar`. You can also pass in a directory that contains the image set .tar file.
+<2> Specify the registry to mirror the image set file to. The registry must start with `docker://`.
++
+// TODO: for <1>: Can there be more than one .tar file in the directory?
```
@afflom @jpower432 For cases where you'd pass in a directory instead of a .tar file, do we expect that there would be more than 1 .tar file in that directory? If so, what would the use case of that be?

bergerhoffer replied:
Thanks, I've updated the description of callout 1 to mention this

bergerhoffer commented on modules/oc-mirror-disk-to-mirror.adoc:
```diff
@@ -0,0 +1,57 @@
+
+* TODO
+
+// TODO: Verification steps - check for manifests. What is the output directory? Wherever you ran the oc mirror command from? Does the oc mirror command create taht directory if it doesn't exist, and create the results-xxx... subfolder. Does the end of the command say which results-xxxx folder it is, just in case there are multiple?
```
@afflom @jpower432 I need a bit of help with what happens after you run this disk2mirror command. If either of you has any example output that you could share, that would help. Also:

* Does it just create the `oc-mirror-workspace` directory in whatever directory the user is in when they run `oc mirror`?
* Does the end of the command give output that shows which results-xxxx folder contains the manifests from this run? (just in case there are multiple)

bergerhoffer commented on modules/oc-mirror-image-set-config-examples.adoc:
```diff
@@ -0,0 +1,31 @@
+TODO
+
+// TODO intro
+// Providing a few examples to accomplish different things. Get help with which ones are most important. Only found one here (https://github.com/openshift/oc-mirror/tree/main/docs/examples) that used storageConfig - I assume we should only document ones that use it?
```
@afflom @jpower432 I was considering adding a section with a few different image set configuration examples. The examples in https://github.com/openshift/oc-mirror/tree/main/docs/examples are all a lot smaller than the example we show in the section on creating the file: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/installing-mirroring-disconnected.html#oc-mirror-creating-image-set-config_installing-mirroring-disconnected

So I assume that means that a lot of that isn't required. So perhaps smaller examples for different use cases might be helpful to users. But I'd be curious to hear whether you think this is worthwhile or not. If not, no worries.

If we do think it's helpful, can either of you help with what examples would be helpful to show. When I look at that list of examples on GitHub, only one of them has "storageConfig". This goes along with my other question on whether "storageConfig" is going to be required. And if so, then I'd think that the examples that we show should all have that.

bergerhoffer commented on modules/oc-mirror-imageset-config-params.adoc:
```diff
@@ -0,0 +1,212 @@
+
+// TODO intro
+
+// TODO: Updated all of the "mappings" to "String" like elsewhere in our doc - okay?
```
@afflom @jpower432 Just as an FYI, I updated all instances of "(mapping)" to stay "String", to be consistent with how we write it elsewhere in the docs. Like here for example: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_aws/installing-aws-customizations.html#installation-configuration-parameters-optional_installing-aws-customizations

I also removed the "Kind" row, because this table is specific to the `ImageSetConfiguration`

bergerhoffer commented on modules/oc-mirror-imageset-config-params.adoc:
```diff
@@ -0,0 +1,212 @@
+|The API version for the `ImageSetConfiguration` content.
+|String. For example: `mirror.openshift.io/v1alpha1`.
+
+// TODO: Changed example (lowered the "v") - confirm
```
@afflom @jpower432 The Google doc had `V1alpha1` as the value for apiVersion, and I updated it to this. Including lowering the "V". Just checking that this is right.

jpower432 replied:
@bergerhoffer Yep! You have it right.

bergerhoffer commented on modules/oc-mirror-imageset-config-params.adoc:
```diff
@@ -0,0 +1,212 @@
+
+[source,yaml]
+----
+additionalImages:
```
@afflom @jpower432 The Google doc had a lowercase "i" in the example here ("additionalimages"). I capitalized it to "additionalImages". Let me know if that is correct.

bergerhoffer commented on modules/oc-mirror-installing-plugin.adoc:
```diff
@@ -0,0 +1,54 @@
+
+.. Navigate to the link:https://console.redhat.com/openshift/downloads[Downloads] page of the {console-redhat-com}.
+
+.. Under the *Command-line interface (CLI) tools* section, click *Download* for *TODO*.
```
@afflom @jpower432 Do we happen to know what the entry for oc-mirror will be called yet for the downloads page in cluster manager?

dmesser replied:
It will likely be called "oc mirror plugin"

bergerhoffer replied:
Thanks, I'll go with this for now, with a note to confirm later

bergerhoffer commented on modules/oc-mirror-installing-plugin.adoc:
```diff
@@ -0,0 +1,54 @@
+.. Under the *Command-line interface (CLI) tools* section, click *Download* for *TODO*.
++
+// TODO: What will it be listed as?
+// TODO: Will they need to select OS/arch?
```
@afflom @jpower432 Also, do we know if they'll have to select OS/arch? And will the file be called just `oc-mirror`, or will there be any extensions or folders?

bergerhoffer replied:
Thanks, we might just have to wait til we get closer to know this info. I'll add reminders to check up on these questions

bergerhoffer commented on modules/oc-mirror-mirror-to-disk.adoc:
```diff
@@ -0,0 +1,61 @@
+
+// TODO: intro
+
+// TODO: usage.md mentioned that this might download hundreds of gigs. And take awhile. Add this warning? And that you can add "log-level=debug" to track progress
```
@afflom @jpower432 Do we need a warning about the potential download size and/or how long it might take to complete?

bergerhoffer commented on modules/oc-mirror-mirror-to-disk.adoc:
```diff
@@ -0,0 +1,61 @@
+* You have installed the OpenShift CLI (`oc`).
+* You have installed the `oc-mirror` CLI plug-in.
+* You have created the image set configuration file.
+// TODO: Don't need a running cluster, but need some pull secrets. Sync w/ team on this
```
@afflom @jpower432 Alex you briefly mentioned that they don't need a running cluster to be able to use oc mirror. But there were some specifics with pull secrets that they did. Can you give me the details on that?

Do they have to do this section (that is in the existing mirroring section), or is it something different: https://docs.openshift.com/container-platform/4.9/installing/installing-mirroring-installation-images.html#installation-adding-registry-pull-secret_installing-mirroring-installation-images

jpower432 replied:
Yes, the configuration of the credentials would be the same except for one difference. I believe the currently mirroring process allows users to pass in a custom file with credentials. `oc-mirror` only supports credentials in the default `docker` location. Our goal is to support the default `podman` location as well before tech preview.

bergerhoffer replied:
I will update the procedure to say to use ~/.docker/config.json. Will ping you all for review once I've updated the PR, to see if what I did was correct or not

bergerhoffer commented on modules/oc-mirror-updating-cluster-manifests.adoc:
```diff
@@ -0,0 +1,36 @@
+
+.Verification
+
+* TODO: Any verification steps?
```
@afflom @jpower432 Are there any verification steps that we can show to help users see that this worked?

afflom reviewed (commented):
This already looks great. Please let me know as you have additional questions.

bergerhoffer commented on modules/installation-adding-registry-pull-secret.adoc:
```diff
@@ -47,8 +54,17 @@ endif::restricted[]
+endif::[]
+ifdef::oc-mirror[]
+. Download your `registry.redhat.io` pull secret from the
+link:https://console.redhat.com/openshift/install/pull-secret[Pull Secret] page on the {console-redhat-com} site and save it to `~/.docker/config.json`.
```
@afflom @jpower432 Can you confirm whether these changes are what we needed to note as the difference in the credentials procedure for oc-mirror? Instead of saving the pull secret in step 1 to an arbitrary .json file, we want to have them save it to `~/.docker/config.json`? 

Though honestly the steps as they are now don't look quite right to me. Step 1 downloads it but as a text file, and then step 3 references some text file that wasn't actually created anywhere in the procedure. Should the real steps be to:

1. Download the pull secret as a text file.
2. Transform that pull secret text file into JSON format. (And for oc-mirror, that should be saved to `~/.docker/config.json`?
3. Then, move forward with getting the username/password (originally step 2)
4. Follow the rest of the steps

Appreciate any thoughts you have on this.

Preview of the section as it is now in the live docs: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-installation-images.html#installation-adding-registry-pull-secret_installing-mirroring-installation-images

Preview of the section for the oc-mirror docs, w/ my updates here: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#installation-adding-registry-pull-secret_installing-mirroring-disconnected

jpower432 replied:
I agree with you there, the steps sound a bit confusing. I agree that steps 2 and 3 should be reversed. We actually also support the default `podman` credential location of `$XDG_RUNTIME_DIR/containers/auth.json` as well

bergerhoffer commented on _topic_maps/_topic_map.yml:
```diff
@@ -95,9 +95,18 @@ Topics:
+    File: installing-mirroring-creating-registry
+  - Name: Mirroring images for a disconnected installation
+    File: installing-mirroring-installation-images
+  - Name: Mirroring images for a disconnected installation using the oc-mirror plug-in
```
@afflom @jpower432 Are the oc-mirror docs appropriate for OKD too, or should it just be for OCP?

jpower432 replied:
Yes, currently the steps for the user would be the same for mirroring OCP or OKD.

bergerhoffer commented on installing/disconnected_install/installing-mirroring-installation-images.adoc:
```diff
@@ -8,12 +8,16 @@ toc::[]
 
 You can use the procedures in this section to ensure your clusters only use container images that satisfy your organizational controls on external content. Before you install a cluster on infrastructure that you provision in a restricted network, you must mirror the required container images into that environment. To mirror container images, you must have a registry for mirroring.
 
+// TODO: Is this procedure going to be marked deprecated for 4.10 so that it could be removed in the future?
```
@afflom @jpower432 Do we have plans to remove the docs for the existing procedures on mirroring in the future? If so, should we mark those docs as deprecated?

jpower432 replied:
I am not aware of any plans for replacing the current docs.

afflom replied:
I'm not sure what the plan is. @dmesser can you assist here?

bergerhoffer replied:
@afflom @dmesser Are there any plans to deprecate/remove the existing disconnected mirroring docs? Or will we leave both that and the oc-mirror options side-by-side going forward?

bergerhoffer commented on installing/disconnected_install/installing-mirroring-disconnected.adoc:
```diff
@@ -0,0 +1,125 @@
+include::modules/oc-mirror-about.adoc[leveloffset=+1]
+
+// TODO: Check the content of this module to make sure it's accurage for oc-mirror
+include::modules/installation-about-mirror-registry.adoc[leveloffset=+1]
```
@afflom @jpower432 I pulled in this module from the existing docs, which talks about the mirror registry. Can you let me know if this makes sense to have in our section as well?

Preview here: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#installation-about-mirror-registry_installing-mirroring-disconnected

jpower432 replied:
I think it does, but we have not verified compatibility with Artificatory, Nexus, or Harbor in the initial release of `oc-mirror`. Our main targets initially were Quay and Docker Registry.

bergerhoffer replied:
Okay so we do also list the registries in the prereqs section here: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#prerequisites_installing-mirroring-disconnected

Should I update these places to only list Quay 3.6+ and Docker Registry? And remove  Artificatory, Nexus, and Harbor?

I'm assuming that the mirror registry for Red Hat OpenShift is still fine though, correct?

afflom replied:
1. mirror registry for openshift is compatible
2. I'd like to get feedback from users trying to integrate other registries with oc-mirror. Can we annotate untested integrations as not supported, but as something that we want feedback on?

bergerhoffer replied:
@afflom thoughts?

bergerhoffer replied:
Ignore my recent comments, I submitted before refreshing ;)

bergerhoffer commented on modules/installation-adding-registry-pull-secret.adoc:
```diff
@@ -48,29 +55,20 @@ Complete the following steps on the installation host:
-----
-<1> For `<user_name>` and `<password>`, specify the user name and password that
-you configured for your registry.
+link:https://console.redhat.com/openshift/install/pull-secret[Pull Secret] page on the {console-redhat-com} site.
```
@jpower432 Okay I updated this file if you can please take another look. I swapped steps 2 and 3, fixed a bit of the wording and step 2 command, and then added a different callout for step 2 for oc-mirror about it having to be one two of those options.

jpower432 replied:
Looks great. Makes much more sense now.

bergerhoffer commented on installing/disconnected_install/installing-mirroring-disconnected.adoc:
```diff
@@ -0,0 +1,124 @@
+
+// TODO: Copied this prereq section + about mirror registry from the orig section. Review/update as necessary
+[id="prerequisites_installing-mirroring-disconnected"]
+== Prerequisites
```
@afflom @jpower432 I pulled in this list of prerequisites (and which supported registry list) from the current docs we have. Can you please review and let me know if it looks correct for our purposes?

The only change I think I made was adding that oc-mirror requires Quay 3.6 or later.

Preview here: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#prerequisites_installing-mirroring-disconnected

jpower432 replied:
Thanks @bergerhoffer, I would say the supported registry list for tech preview would be Quay (3.6 or later) and Docker Registry. @afflom Do you agree? Would it be worth it to verify the others right now?

afflom replied:
I agree with CNCF Distribution and Quay 3.6+. We've also tested go-containerregistry pretty extensively with oc-mirror without issue. Not sure if thats something that we want to advertise.

bergerhoffer replied:
@jpower432 @afflom So can you sync up and confirm which ones we want to list? I see consensus on Quay 3.6+, but I'm not sure if Docker Registry and CNCF are the same or different, etc.

afflom replied:
Docker registry was donated to CNCF. I think the names are interchangeable. Docker registry and quay 3.6+ are definitely tested. We need to test the others listed.  We dont feel comfortable mentioning them as compatible yet. We do know that the 4.10 release of oc-mirror does not work with Harbor. Can we tag artifactory and nexus as untested instead of unsupported/omitted so that we dont discourage people from testing them?

bergerhoffer commented on modules/oc-mirror-image-set-config-examples.adoc:
```diff
@@ -0,0 +1,46 @@
+== Include arbitrary images and helm charts
+
+The following `ImageSetConfiguration` file includes a UBI image and helm charts.
+// TODO: get more specific?
```
@afflom @jpower432 How does this first example file look, and is this description okay?

afflom replied:
The example looks great. Only nit is that the example description does not include a reference to the `additionalImages` configuration (it only references helm).

bergerhoffer replied:
Thanks, I updated the descriptions a bit after I added the second example. PTAL to and let me know what you think now

bergerhoffer commented on modules/oc-mirror-image-set-config-examples.adoc:
```diff
@@ -0,0 +1,46 @@
+----
+// TODO: Update/strip example as needed
+
+// TODO: Add another example
```
@afflom @jpower432 Can you help me with another good example image set config file that we can show?

I updated the main image set config creation procedure to have a pretty stripped down example: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#oc-mirror-creating-image-set-config_installing-mirroring-disconnected

And then the first example here that I have adds the helm charts and additional images: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#oc-mirror-image-set-examples_installing-mirroring-disconnected

Most of the examples in the repo (https://github.com/openshift/oc-mirror/tree/main/docs/examples) don't have a storage back end in the config. So I could use help with what would be a good second example to show.

bergerhoffer replied:
Thanks @jpower432. Does it matter where the storageConfig is located? In the above example it's at the bottom, so I'd like to know which is the best way and be consistent

jpower432 replied:
@bergerhoffer I prefer to keep it at the top because of how long the mirror spec can get, but it will work either way.

bergerhoffer replied:
Thanks, I'll move the other example to have it at the top, for consistency

bergerhoffer commented on modules/oc-mirror-updating-cluster-manifests.adoc:
```diff
@@ -0,0 +1,47 @@
+
+After you have mirrored your image set to the mirror registry, you must install the generated `ImageContentSourcePolicy` and `CatalogSource` resources into the cluster.
+
+// TODO: why?
```
@afflom @jpower432 Can you explain in simple terms what exactly this is doing / why this is needed? The google doc says "oc-mirror will output kubernetes manifests that are required by OpenShift for nodes to utilize the images within the mirror registry. " but it's still not clear to me.

jpower432 replied:
Sure! Since `oc-mirror` is tracking which images are being requested originally and where they are published in the target mirror registry, it creates `ImageContentSourcePolicy` (ICSP) custom resources to define the mirror location for the original images. This is especially helpful when it comes to utilizing the custom catalog images being built. `oc-mirror` generates a `CatalogSource` to allow the catalog to be utilized by the cluster; the operators images being referenced by the catalogs are still defined as their original image location. The ICSP allows the source and mirrored location to be checked for the operator container image. Does this help? Perhaps it might be useful to link to ICSP documentation. WDYT?

bergerhoffer replied:
Thanks @jpower432 that helps. I took a look at elsewhere that we talk about ICSPs and catalog sources and took a stab at a description for them. Can you take a look at what I say now and correct anything that's not right?

```
 The `ImageContentSourcePolicy` resource associates the mirror registry with the source registry and redirects container pull requests from the online registries to the mirror registry. The `CatalogSource` resource is used by OLM to retrieve information about the available Operators in the mirror registry.
```

Preview: https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#oc-mirror-updating-cluster-manifests_installing-mirroring-disconnected

jpower432 replied:
@bergerhoffer One nit here:  I would replace `container` with `image`.

sagidlow commented on modules/oc-mirror-installing-plugin.adoc:
```diff
@@ -0,0 +1,53 @@
+[id="installation-oc-mirror-installing-plugin_{context}"]
+= Installing the oc-mirror OpenShift CLI plug-in
+
+In order to use the oc-mirror OpenShift CLI plug-in to mirror registry images, you must install the plug-in. If you are mirroring image sets in a fully disconnected environment, ensure that you install the oc-mirror plug-in on all required hosts.
```
```suggestion
To use the oc-mirror OpenShift CLI plug-in to mirror registry images, you must install the plug-in. If you are mirroring image sets in a fully disconnected environment, ensure that you install the oc-mirror plug-in on all required hosts.
```
Minimalism.

sagidlow commented on modules/oc-mirror-mirror-to-mirror.adoc:
```diff
@@ -0,0 +1,53 @@
+The initial image set to populate the mirror registry is often the largest. Subsequent executions of oc-mirror only download the changes since the previous execution, so the generated image set is often a smaller size.
+====
+
+You are required to specify a storage back end in the image set configuration file. This storage back end can be a local directory or a docker v2 registry. The oc-mirror plug-in stores metadata in this storage back end during image set creation.
```
Should Docker be capitalized?

sagidlow reviewed (commented):
Lots of good work here. Just a couple of nit comments. Otherwise, LGTM.

bergerhoffer commented on installing/disconnected_install/installing-mirroring-creating-registry.adoc:
```diff
@@ -0,0 +1,43 @@
++
+[IMPORTANT]
+====
+These requirements are based on local testing results with only Release images and Operator images tested. Storage requirements can vary based on your organization's needs. Some users might require more space, for example, when they mirror multiple z-streams. You can use standard Red Hat Quay functionality to remove unnecessary images and free up space.
```
@stevsmit this content was copied from your mirror registry files. I can see that "Release images" is capitalized a few times in the file, can you let me know if they do really need to be capitalized, or can they be lowered?

stevsmit replied:
I actually did this consciously because the dev team + Daniel made it look like it should always be capitalized (as in, any time they wrote Release images they capitalized it). That said, I just talked to our Principal dev and he thinks it should be lower case. I agree; I only ever did it because higher-stakes people _always_ capitalized it.

bergerhoffer commented on installing/disconnected_install/installing-mirroring-creating-registry.adoc:
```diff
@@ -0,0 +1,43 @@
+* Key-based SSH connectivity on the target host. SSH keys are automatically generated for local installs. For remote hosts, you must generate your own SSH keys.
+* 2 or more vCPUs.
+* 8 GB of RAM.
+* About 12 GB for {product-title} 4.10 Release images, or about 358 GB for {product-title} 4.10 Release images and {product-title} 4.10 Red Hat Operator images. Up to 1 TB per stream or more is suggested.
```
@stevsmit Can these 4.10's just be removed, or is the version actually important to show here?

jpower432 commented on modules/oc-mirror-creating-image-set-config.adoc:
```diff
@@ -0,0 +1,46 @@
+
+[IMPORTANT]
+====
+Do not delete or modify the metadata that is generated by the oc-mirror plug-in executions. You must use the same storage back end on subsequent executions of oc-mirror to the same mirror registry.
```
@bergerhoffer In this message about the metadata or maybe in a separate IMPORTANT message, we need to alert users that if they specify a top-level namespace in their initial run of `oc-mirror` they cannot change it on subsequent runs.
Ex.
```
./bin/oc-mirror --config ~/configs/imageset-config.yaml docker://localhost:5000/mynamespace
```
`mynamespace` cannot change

bergerhoffer replied:
Thanks @jpower432! I make this mention a few places if you can see what you think. Diff of my changes is here: https://github.com/openshift/openshift-docs/compare/b726ac339e22ff9dc29e653cafc280d54ecb7a50..79be0b985d976af7506281c84045ecb3ef73c7dc

jpower432 replied:
Looks good. Thanks!

bergerhoffer commented:
@zhouying7780 This is still going through SME review, but please start taking a look these oc-mirror docs whenever you get a chance. 

More details are in this comment in Jira: https://issues.redhat.com/browse/OSDOCS-2891?focusedCommentId=19817600&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-19817600

Preview links are in the first comment of the PR.

afflom commented on modules/oc-mirror-imageset-config-params.adoc:
```diff
@@ -0,0 +1,208 @@
+|The Operator catalog to include in the image set.
+|String. For example: `registry.redhat.io/redhat/redhat-operator-index:v4.9`.
+
+|`mirror.operators.headsOnly`
```
This boolean cannot be used if the packages subkey is present. Can you please add a note about that here?

afflom commented on modules/oc-mirror-about.adoc:
```diff
@@ -0,0 +1,23 @@
+* Provides a centralized method to mirror {product-title} releases, Operators, helm charts, and custom images.
+* Maintains update paths for {product-title} and Operators.
+* Uses a declarative image set configuration file to include only the {product-title} releases, Operators, and images that your cluster needs.
+* Performs incremental mirroring, which reduces the size of future image sets.
```
I think that oc-mirror update operations are differential vice incremental, but that might be more of a nit.

bergerhoffer replied:
I had actually pulled that from the slide of the what's new presentation where Daniel talked about disconnected mirroring: https://speakerdeck.com/redhatlivestreaming/whats-new-in-openshift-4-dot-10?slide=10

Let me know if I should change it here.

afflom commented on modules/oc-mirror-about.adoc:
```diff
@@ -0,0 +1,23 @@
+
+The oc-mirror OpenShift CLI (`oc`) plug-in allows you to mirror all required {product-title} content and custom images to your mirror registry by using a single tool. It provides the following features:
+
+* Provides a centralized method to mirror {product-title} releases, Operators, helm charts, and custom images.
```
the additionalImages field can pull official/vendor images along with custom images. Can this be changed to `other images` or `arbitrary images` or something else?

afflom reviewed (changes requested):
Wow, this is really amazing work! May all of the instances of the words "image set" be combined as "imageset" and all of the instances of the words "back end" be combined as "backend"?

bergerhoffer commented:
@afflom Per our slack convo, I updated to use "backend" one word, but we're keeping "image set" two words.

I also updated the list of supported registries to just Quay, Docker Registry, and the mirror registry, if you can please check the two places we mention to make sure it's good:

* https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#prerequisites_installing-mirroring-disconnected
* https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#installation-about-mirror-registry_installing-mirroring-disconnected

afflom commented:
> @afflom Per our slack convo, I updated to use "backend" one word, but we're keeping "image set" two words.
> 
> I also updated the list of supported registries to just Quay, Docker Registry, and the mirror registry, if you can please check the two places we mention to make sure it's good:
> 
> * https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#prerequisites_installing-mirroring-disconnected
> * https://deploy-preview-40474--osdocs.netlify.app/openshift-enterprise/latest/installing/disconnected_install/installing-mirroring-disconnected.html#installation-about-mirror-registry_installing-mirroring-disconnected

Looks great! Thank you

zhouying7780 commented on modules/oc-mirror-installing-plugin.adoc:
```diff
@@ -0,0 +1,56 @@
+
+.. Navigate to the link:https://console.redhat.com/openshift/downloads[Downloads] page of the {console-redhat-com}.
+
+.. Under the *Command-line interface (CLI) tools* section, click *Download* for *OpenShift Client (oc) mirror plugin* and save the file.
```
From now , could not see the download for oc plugin

bergerhoffer replied:
Right, this isn't available yet on this page - I assume it will not be there until 4.10 GA.

@afflom @jpower432 Can you help @zhouying7780 with if there is a way to test this now, without there being the binary available on the live downloads page yet?

afflom replied:
Sorry that I missed this tag. I've posted a one liner for Ying in slack to retrieve the oc-mirror from the release payload.

zhouying7780 commented on modules/oc-mirror-mirror-to-disk.adoc:
```diff
@@ -0,0 +1,69 @@
+====
+
+You are required to specify a storage backend in the image set configuration file. This storage backend can be a local directory or a docker v2 registry. The oc-mirror plug-in stores metadata in this storage backend during image set creation.
+
```
For the 'fully disconnected environment', we only could sepcify the storage backend  for local directory, can't be a docker registry.

bergerhoffer replied:
@zhouying7780 please see @afflom's comment below about the registry option still being available to use here.

afflom commented:
Sorry to jump in here, but we encourage users to always use a registry
backend for mirror2disk and mirror2mirror workflows; even when creating an
imageset intended to be used fully internet-disconnected. If they are
unable to use a registry backend, something within their environment is
misconfigured or there is a bug within oc-mirror.

On Sun, Feb 27, 2022, 20:29 Ying Zhou ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In modules/oc-mirror-mirror-to-disk.adoc
> <https://github.com/openshift/openshift-docs/pull/40474#discussion_r815557745>
> :
>
> > +
> +:_content-type: PROCEDURE
> +[id="oc-mirror-mirror-to-disk_{context}"]
> += Mirroring from mirror to disk
> +
> +You can use the oc-mirror plug-in to generate an image set and save the contents to disk. The generated image set can then be transferred to the disconnected environment and mirrored to the target registry.
> +
> +[IMPORTANT]
> +====
> +Depending on the configuration specified in the image set configuration file, using oc-mirror to mirror images might download several hundreds of gigabytes of data to disk.
> +
> +The initial image set to populate the mirror registry is often the largest. Subsequent executions of oc-mirror only download the changes since the previous execution, so the generated image set is often a smaller size.
> +====
> +
> +You are required to specify a storage backend in the image set configuration file. This storage backend can be a local directory or a docker v2 registry. The oc-mirror plug-in stores metadata in this storage backend during image set creation.
> +
>
> For the 'fully disconnected environment', we only could sepcify the
> storage backend for local directory, can't be a docker registry.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/openshift-docs/pull/40474#pullrequestreview-894623224>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHKBMT4BD257MPUP4Q74HSTU5LTZDANCNFSM5LXG7BMA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

zhouying7780 commented:
Look good to me , thank you !

kalexand-rh commented on modules/oc-mirror-updating-cluster-manifests.adoc:
```diff
@@ -0,0 +1,47 @@
+
+. Log in to the OpenShift CLI as a user with the `cluster-admin` role.
+
+. Apply the generated manifests to the cluster:
```
Ok, I see where `results-1639608409` came from, but everywhere else, you call it a results directory, not manifests. It might be easier to see what's going on if you "apply the contents of the results directory" or something similar here or talk about the manifests in the results folder elsewhere.

kalexand-rh commented on modules/oc-mirror-updating-cluster-manifests.adoc:
```diff
@@ -0,0 +1,47 @@
+$ oc apply -f ./oc-mirror-workspace/results-1639608409/
+----
+
+// TODO: Any example output to show?
```
++

kalexand-rh reviewed (commented):
I only looked at the raw files, but this is looking good! I have a few suggestions and some picky style stuff, but this is so close.

zhouying7780 commented on modules/oc-mirror-disk-to-mirror.adoc:
```diff
@@ -0,0 +1,43 @@
+[source,terminal]
+----
+$ oc mirror --from=./mirror_seq1_000000.tar \ <1>
+  docker://registry.example:5000/             <2>
```
the docker url should be: "docker://registry.example:5000"  or "docker://registry.example:5000/namespace",  if no namespace should remove '/'

bergerhoffer replied:
Thanks, updated

zhouying7780 commented on modules/oc-mirror-mirror-to-mirror.adoc:
```diff
@@ -0,0 +1,53 @@
+[source,terminal]
+----
+$ oc mirror --config=./imageset-config.yaml \ <1>
+  docker://registry.example:5000/             <2>
```
some here , the docker url should be : "docker://registry.example:5000"  or "docker://registry.example:5000/namespace" , if no namespace should remove '/'

bergerhoffer replied:
Updated!

zhouying7780 commented on modules/oc-mirror-mirror-to-disk.adoc:
```diff
@@ -0,0 +1,69 @@
+[source,terminal]
+----
+$ oc mirror --config=./imageset-config.yaml \ <1>
+  file:///<path_to_output_directory>          <2>
```
should be : file://<path_to_output_directory>

zhouying7780 replied:
Not another `/` is needed , we need to remove one `/`  from 'file:///'

zhouying7780 commented:
Have raise some comments, other look good to me.

bergerhoffer commented on installing/disconnected_install/installing-mirroring-creating-registry.adoc:
```diff
@@ -0,0 +1,45 @@
+* Key-based SSH connectivity on the target host. SSH keys are automatically generated for local installs. For remote hosts, you must generate your own SSH keys.
+* 2 or more vCPUs.
+* 8 GB of RAM.
+* About 12 GB for {product-title} {product-version} release images, or about 358 GB for {product-title} {product-version} release images and {product-title} {product-version} Red Hat Operator images. Up to 1 TB per stream or more is suggested.
```
Same for this one FYI @stevsmit

bergerhoffer commented on modules/installation-about-mirror-registry.adoc:
```diff
@@ -1,16 +1,27 @@
 You can use any container registry that supports link:https://docs.docker.com/registry/spec/manifest-v2-2[Docker v2-2], such as Red Hat Quay, the _mirror registry for Red Hat OpenShift_, Artifactory, Sonatype Nexus Repository, or Harbor. Regardless of your chosen registry, the procedure to mirror content from Red Hat hosted sites on the internet to an isolated image registry is the same. After yo ...
+endif::[]
+ifdef::oc-mirror[]
+You can mirror the images that are required for {product-title} installation and subsequent product updates to a container mirror registry such as Red Hat Quay 3.6 or later or Docker Registry. If you do not have access to a large-scale container registry, you can use the _mirror registry for Red Hat OpenShift_, which is a small-scale container registry included with {product-title} subscriptions.
```
Thanks @kalexand-rh I wasn't aware of that if it's true.

@sferich888 do you have thoughts on this?

sferich888 replied:
Our testing (for this feature) or the scope of what it should work with should match our product registry requirements. For that any mirroring should match these requirments: [https://docs.openshift.com/container-platform/4.9/installing/installing-mirroring-installation-images.html#prerequisites_insta[…]installation-images](https://docs.openshift.com/container-platform/4.9/installing/installing-mirroring-installation-images.html#prerequisites_installing-mirroring-installation-images)

IE: container image registry that supports [Docker v2-2](https://docs.docker.com/registry/spec/manifest-v2-2)

Thus; I don't think we need to call out any registries (we do that already to list 'possibilities' in other parts of the docs; we should only call out this requirement.

IE: the mirror tool should work with ANY registry that offers/supports Docker v2-2 as a protocol.

bergerhoffer replied:
Per slack convo with Eric, he recommended that we avoid naming specific registries that we've tested. And instead focus on saying that we support any registry that supports docker v2-2, even if we haven't specifically tested all vendors. Making an update to the oc-mirror docs for this, but will leave the existing docs that had all registries listed alone

bergerhoffer commented on installing/disconnected_install/installing-mirroring-disconnected.adoc:
```diff
@@ -0,0 +1,133 @@
+* Use cases for changing image set config between runs
+////
+
+You can use the procedures in this section to ensure your clusters only use container images that satisfy your organizational controls on external content. Before you install a cluster on infrastructure that you provision in a restricted network, you must mirror the required container images into that environment. To mirror container images, you must have a registry for mirroring.
```
Thanks, will update this in the assembly I pulled it from too!

bergerhoffer commented on installing/disconnected_install/installing-mirroring-disconnected.adoc:
```diff
@@ -0,0 +1,133 @@
+
+You can use the procedures in this section to ensure your clusters only use container images that satisfy your organizational controls on external content. Before you install a cluster on infrastructure that you provision in a restricted network, you must mirror the required container images into that environment. To mirror container images, you must have a registry for mirroring.
+
+You can use the oc-mirror OpenShift CLI (`oc`) plug-in to mirror images to a mirror registry in your fully or partially disconnected environments.
```
Yeah I know, I went back and forth on this one for awhile. It's kind of a gray area, and it was too overwhelming to mark it up

bergerhoffer commented on modules/oc-mirror-creating-image-set-config.adoc:
```diff
@@ -0,0 +1,46 @@
++
+This example will pull images from the `stable-4.9` channel for the `registry.redhat.io/redhat/redhat-operator-index:v4.9` operator catalog and saves the image set metadata to the `example.com/example/oc-mirror` registry.
+
+. Save the file as `imageset-config.yaml`. This file is required by the `oc mirror` command when mirroring content.
```
+1 for it being a bit confusing to document and read. If I had my preference, it would just be an `oc mirror` command that was already part of the `oc` binary. But I'll bring it up to the team for discussion prior to GA, just in case they want to consider any changes.

bergerhoffer commented on installing/disconnected_install/installing-mirroring-disconnected.adoc:
```diff
@@ -0,0 +1,133 @@
+
+// TODO: Copied this prereq section + about mirror registry from the orig section. Review/update as necessary
+[id="prerequisites_installing-mirroring-disconnected"]
+== Prerequisites
```
I see what you mean. At first, the placement made sense to me since it follows after two "About" modules, so that you learn about the oc-mirror tool and about the mirror registry. Though this might be a chicken and egg thing, where why do you need to learn about the mirror registry til you know it's a prereq? 

Honestly though, the only prereq listed is the mirror registry. I might move the prereqs above the mirror registry module for now, but I'll re-evaluate it for the GA release to see if they should be combined, so it's not so redundant.

bergerhoffer commented on modules/oc-mirror-disk-to-mirror.adoc:
```diff
@@ -0,0 +1,43 @@
+  docker://registry.example:5000/             <2>
+----
+<1> Pass in the image set .tar file to mirror. This procedure assumes that it is named `mirror_seq1_000000.tar`. If an `archiveSize` value was set in the image set configuration file, the image set might be broken up into multiple .tar files. In this situation, you can pass in a directory that contains the image set .tar files.
+<2> Specify the registry to mirror the image set file to. The registry must start with `docker://`. If you specify a top-level namespace for the mirror registry, you must also use this same namespace on subsequent executions.
```
Agree. It's on my todo list that I just didn't get to for this TP release, but plan to add for the GA release next time.

bergerhoffer commented on modules/oc-mirror-mirror-to-disk.adoc:
```diff
@@ -0,0 +1,69 @@
+
+.Next steps
+
+* Transfer the image set .tar file to the disconnected environment.
```
You know, I was uncertain on this as well so I checked the ISG on this, and they didn't have it marked up:

![image](https://user-images.githubusercontent.com/7443887/156394682-a9ab8143-54b6-473f-9903-9a57821680bd.png)

bergerhoffer commented on modules/oc-mirror-imageset-config-params.adoc:
```diff
@@ -0,0 +1,208 @@
+|String. For example: `registry.redhat.io/ubi8/ubi:latest`
+
+|`mirror.helm`
+|The helm configuration of the image set. Note that the oc-mirror plug-in only supports helm charts that do not require user input when rendered.
```
Not familiar enough with helm charts to know. @afflom Can you confirm if "rendered" is the appropriate term here?

afflom replied:
Rendering or rendered. Either seems fine to me.

bergerhoffer commented on modules/oc-mirror-imageset-config-params.adoc:
```diff
@@ -0,0 +1,208 @@
+
+|`mirror`
+|The configuration of the image set.
+|Object
```
Not in time for 4.10, but I'll add it as a todo for next release

bergerhoffer commented on modules/oc-mirror-creating-image-set-config.adoc:
```diff
@@ -0,0 +1,46 @@
+
+[IMPORTANT]
+====
+Do not delete or modify the metadata that is generated by the oc-mirror plug-in executions. You must use the same storage backend on subsequent executions of oc-mirror to the same mirror registry.
```
Yikes, this is the wording I asked you for help with the other day, then forgot to actually make the updates :grimacing:

zhouying7780 commented:
/lgtm

bergerhoffer commented on modules/oc-mirror-disk-to-mirror.adoc:
```diff
@@ -0,0 +1,43 @@
+<1> Pass in the image set .tar file to mirror. This procedure assumes that it is named `mirror_seq1_000000.tar`. If an `archiveSize` value was set in the image set configuration file, the image set might be broken up into multiple .tar files. In this situation, you can pass in a directory that contains the image set .tar files.
+<2> Specify the registry to mirror the image set file to. The registry must start with `docker://`. If you specify a top-level namespace for the mirror registry, you must also use this same namespace on subsequent executions.
+
+This command updates the mirror registry with the image set and generates the `ImageContentSourcePolicy` and `CatalogSource` resources.
```
Good catch, thanks

bergerhoffer commented on modules/oc-mirror-imageset-config-params.adoc:
```diff
@@ -0,0 +1,208 @@
+
+The following table lists the available parameters for the `ImageSetConfiguration` resource.
+
+.`ImageSetConfiguration` parameters
```
Yeah I know what you mean. Though irrespective of time, I've usually avoided doing more than 3 columns due to how tables render poorly on the portal. Not sure if this has been improved since the last time I've checked.

bergerhoffer commented on modules/oc-mirror-installing-plugin.adoc:
```diff
@@ -0,0 +1,56 @@
+[id="installation-oc-mirror-installing-plugin_{context}"]
+= Installing the oc-mirror OpenShift CLI plug-in
+
+To use the oc-mirror OpenShift CLI plug-in to mirror registry images, you must install the plug-in. If you are mirroring image sets in a fully disconnected environment, ensure that you install the oc-mirror plug-in on all required hosts.
```
It would be the host that has access to the internal, and then whatever host you're on in the disconnected env that has access to the mirror registry. I'll play with it to see if I can clarify

bergerhoffer commented on modules/oc-mirror-differential-updates.adoc:
```diff
@@ -0,0 +1,40 @@
+
+.Procedure
+
+. Follow the same steps that you used to create the initial image set and mirror it to the mirror registry.
```
That remains to be seen :wink: At the very least, I'll add the procedures as additional resources for this module, and see if there's anything else I can clarify here.

afflom commented:
/lgtm

bergerhoffer commented:
/cherrypick enterprise-4.10
